### PR TITLE
Add platform logging and view

### LINKED ISSUE #518: Audit logging of admin events
### Epic

_No response_

### Description

**As a:** admin

**I want to:** see platform wide events in the audit log in #517 such as:
 - user creation
 - user deletion
 - password resets
 - team creation
 - team deletion

**So that:** I can observe activity across the platform and Identify when users are created

### Acceptance Criteria

- [ ] criteria 1
- [ ] criteria 2
- [ ] criteria 3
- ...

### PR DESCRIPTION
**Pull request by Steve-Mcl**
fixes #518

1st Draft - so progress and directions can be evaluated.

### NOTES...

* Uses `entityType: 'platform'`  and `entityId: 'audit'` 
  *  `entityId` of "audit" is a placeholder until something more suitible is decided
* Reuses AuditLog from '@/components/AuditLog' for the view
* adds permission `platform:audit-log`
* logs (so far)...
   * platform.licence.apply
   * platform.licence.inspect
   * platform.settings.update

### Screenshot...
image

### TODO
* logging of items listed in #518
* tests

### CODE REVIEW

**Comment by Steve-Mcl** [comment]
@knolleary - would appreciate once over in case this is not the direction intended.

Ta.

**Review by knolleary — CHANGES_REQUESTED** [review]
You can simplify the access control for the admin audit-log view - comments inline.

We should be mindful of what data gets included with the audit log entries - it currently stores all old/new values. We will need to make sure things like passwords *never* get accidentally stored that way for example.

@joepavitt has been doing design work around the existing audit logs. Please talk to him to make sure this will fit with what he's got in #428.

**Comment by Steve-Mcl** [comment]
@knolleary 


* As the `entityType` is "platform", does the `event` need to be prefixed with "plaform."?
* For now, I have entered "audit" as the `entityId` because this is not a "Team" or a "Project" but a platform "thing". Any preference on what goes into `entityId`?


image

**Comment by Steve-Mcl** [comment]
> @joepavitt has been doing design work around the existing audit logs. Please talk to him to make sure this will fit with what he's got in https://github.com/flowforge/flowforge/issues/428.

It should since I re-use the component but will need to check the JSON output and if Joe is expecting or hoping to to implement a well defined format (e.g. a `level` property for colourisation? a well known property name for friendly text, a particular property for `details` )

**Comment by Steve-Mcl** [comment]
> For `platform.settings.update` you have the example body being related to users signing up? Not sure the event name and body align there?

No, its Admin → settings → general → Allow new users to register on the login screen
The original value was `false` and was changed to `true`
image
resulting in...
image


For example, enabling T+Cs & setting URL in the same admin settings page
image
 results in ...
image
... where original and new settings are logged.

---

> I'll come up with a "License" icon, and then map platform.settings.update once I have a clearer idea on what that entails.

There are many more being added. For example (not definitive - I am still working through them all)...
```log
platform.licence.apply
platform.licence.inspect

platform.users.create-user
platform.users.auto-create-team
platform.users.delete
platform.users.update-user

platform.account.login
platform.account.logout
platform.account.register
platform.account.verify-email-token
platform.account.auto-create-team
platform.account.verify
platform.account.forgot-password
platform.account.reset-password
```

---

For the majority of these new platform level audit log entries...
* when successful, the `body` is `{"some": "thing"}`  (_no standard as of writing_) 
**OR** 
* when they fail, the `body` is per the new API response pattern e.g. `{ code: 'not_enabled', error: 'Email not enabled - cannot reset password' }`  plus some related properties (like the value they tried to enter)

Personally, I think we should consider a formal pattern for audit logging.  Perhaps (as previously mentioned) something like...

##### Success Log Entry
```
{ level: "info", info: { "some": "detailed json based info with useful info", "name": "Steve", "age": "old" }
```


##### Error Log Entry
```
{ level: "error":, code: "something_meaningful_to_system", error: "human friendly error", detail: { "some": "detailed json based info with useful info", "lottery-win": "no", "rich": false }
```


With some structure, we could auto determine icon (or default) & colourisation for new and unlisted log entries. 

Thoughts please @joepavitt @knolleary

**Comment by joepavitt** [comment]
> Personally, I think we should consider a formal pattern for audit logging

I am all for standardisation of object classes/structure - so you have my support.

We seem to have two classifications of a logging event, where I currently capture just one with the iconography:
- Scope: (Platform/Team/Project/Node-RED?)
- Type: Settings, License, Account, Device, Snapshot, Billing, etc.

Iconography, in most cases just relies on the type at the moment, but I may need to revisit this.

**Comment by Steve-Mcl** [comment]
### Notes.

Lots of progress - felt it was time the draft was updated to ensure this is all still heading in the right direction.

* I have modified the `entity` for the user centric stuff to `"user"` log as discussed.
* The `entityId` for `"platform"` log entries will be `null` as requested
* Typically the `entityId` for `"user"` log entries will be the user ID of the affected user (not the actioning user)
  * currently, this is not returned in the view
* Some of the more common audit events added have been covered in existing tests or new tests
  * there is a question about whether the whole audit logging even needs tests or if they should be a single test file?
* each error should now contain a `code` and `error` property - this could be used for keying or colourisation?
* each success should now contain a `status` of "okay" again, this could be used for keying or colourisation?
* There are likely more but as there is no acceptance criteria I am having to hunt down the gaps
* not all failure code paths have been tested (neither manually nor via tests) - BUT - so far what I have tested works as expected.



### platform log
event name | associated resp code | sample data in `body`
-- | -- | --
platform.settings.update  |  200  |  `{ status: 'okay', changes: { [key]: { old:xxx, new:yyy }} }`
platform.licence.apply  |  200  |  `{ status: 'okay', license: {} }`
platform.licence.apply  |  400  |  `{ code: 'invalid_license', error: responseMessage }`
platform.licence.inspect  |  200  |  `{ status: 'okay', license: {} }`
platform.licence.inspect  |  400  |  `{ code: 'invalid_license', error: responseMessage }`


### user log - account
event name | associated resp code | sample data in `body`
-- | -- | --
account.login  |  200  |  `{ status: 'okay', user: { username: '' } }`
account.login  |  401  |  `{ code: 'unauthorized', error: 'unauthorized', user: { username: '' } }`
account.login  |  403  |  `{ code: 'user_suspended', error: 'User Suspended', user: { username: '' } }`
account.logout  |  200  |  `{ status: 'okay', user: { id: '' } }`
account.register  |  200  |  `{ status: 'okay', user: {...} }`
account.register  |  400  |  `{ code: 'not_enabled', error: 'user registration not enabled', user: {...} } `
account.register  |  400  |  `{ code: 'invalid_request', error: 'user registration not enabled - email not configured', user: {...} } `
account.register  |  400  |  `{ code: 'invalid_request', error: 'invalid username', user: {...} } `
account.register  |  400  |  `{ code: 'invalid_request', error: 'terms and conditions not accepted', user: {...} } `
account.register  |  400  |  `{ code: 'unexpected_error' | 'not_available', error: 'error info here', user: {...} }`
account.verify.request-token  |  200  |  `{ status: 'okay', user: {...} }`
account.verify.request-token  |  400  |  `{ code: 'invalid_request', error: 'email not configured' }`
account.verify.request-token  |  400  |  `{ code: 'invalid_request', error: 'email already verified' }`
account.verify.verify-token  |  200  |  `{ status: 'okay', user: {...} }`
account.verify.verify-token  |  400  |  `{ code: 'team_limit_reached', error: 'Unable to create user team: license limit reached' }`
account.verify.verify-token  |  400  |  `{ code: 'invalid_request', error: err.toString() }`
account.verify.auto-create-team  |  xxx  |  `{ status: 'okay', team: { name: '', type: 'starter'}`
account.reset-password  |  200  |  `{ userId, status: 'okay' }`
account.reset-password  |  400  |  `{ code: 'not_enabled', error: 'password reset not enabled' }`
account.reset-password  |  400  |  `{ code: 'unexpected_error', error: 'Password reset failed' }`
account.forgot-password  |  200  |  `{ status: 'okay', info: 'Password reset request for ${user.hashid}' }`
account.forgot-password  |  400  |  `{ code: 'not_enabled', error: 'password reset not enabled', user: { email: '' } }`
account.forgot-password  |  400  |  `{ code: 'not_enabled', error: 'Email not enabled - cannot reset password' }`


### user log - users
event name | associated resp code | sample data in `body`
-- | -- | --
users.create-user  |  200  |  `{ status: 'okay', user: {...} }`
users.create-user  |  400  |  `{ code: 'invalid_username', error: 'invalid username', user: {...} }`
users.create-user  |  400  |  `{ code: 'team_limit_reached', error: 'Unable to create user team: license limit reached', user: {...} }`
users.create-user  |  400  |  `{ code: 'unexpected_error' | 'invalid_username' | 'invalid_email', error: 'error info here', user: {...} }`
users.auto-create-team  |  200  |  `{ status: 'okay', team: { name: `Team ${request.body.name}`, type: 'starter' }, user: {...} }`
users.delete  |  200  |  `{ status: 'okay', user: {...} }`
users.delete  |  400  |  `{ code: 'unexpected_error', error: 'error info here', user: {...} }`
users.update-user  |  200  |  `{ status: 'okay', old: originalValues, new: newValues, user: { name, username, email } }`
users.update-user  |  400  |  `{ code: 'invalid_request', error: 'cannot suspend self', user: { name, username, email } }`
users.update-user  |  400  |  `{ code: 'invalid_team', error: 'invalid team', team: request.body.defaultTeam, user: { name, username, email } }`
users.update-user  |  400  |  `{ code: 'unexpected_error', error: responseMessage, user: { name, username, email } }`


### user log - user
event name | associated resp code | sample data in `body`
-- | -- | --
user.change-password  |  200  |  `{ status: 'okay' }`
user.change-password  |  400  |  `{ code: 'password_change_failed', error: 'password change failed' }`
user.update-user  |  200  |  `{ status: 'okay', old: originalValues, new: newValues, user: { name, username, email } }`
user.update-user  |  400  |  `{ code: 'invalid_request', error: 'cannot suspend self', user: { name, username, email } }`
user.update-user  |  400  |  `{ code: 'invalid_team', error: 'invalid team', team: request.body.defaultTeam, user: { name, username, email } }`
user.update-user  |  400  |  `{ code: 'unexpected_error', error: responseMessage, user: { name, username, email } }`


### user log - user invitation
event name | associated resp code | sample data in `body`
-- | -- | --
user.invitations.accept-invite  |  200  |  `{ status: 'okay' }`
user.invitations.delete-invite  |  200  |  `{ status: 'okay' }`

**Comment by knolleary** [comment]
As I said in our conversation on Monday, I was not in favour of spending time now reworking the format of all of the audit log entries as that was out of scope of the requirement and would require additional time to get right. I'm not saying its a bad thing to do - it is certainly worthwhile - but it has added days of additional work to this and I'm not clear how much more work still needs to be done.

> Each success should now contain a status of "okay" again

That is redundant. If any failure will have a `code`/`error` property, then we don't need to add anything for success. Please remove.

> Typically the entityId for "user" log entries will be the user ID of the affected user (not the actioning user)
currently, this is not returned in the view

Can we be more specific about what 'typically' means here? Where are there exceptions to this?

I can see there being two cases:
 - a user is updating their own account
 - an admin is updating another users account

How do you currently distinguish between the?

**Comment by Steve-Mcl** [comment]
> > Each success should now contain a status of "okay" again
> 
> That is redundant. If any failure will have a `code`/`error` property, then we don't need to add anything for success. Please remove.

Are you sure you want me to remove this? I aligned with the API reply and many (most) of them reply `{status: 'okay'}`
I simply included it as there can be value (i.e. something can complete but not be an error and not be ok).

**Comment by Steve-Mcl** [comment]
> > Typically the `entityId` for "user" log entries will be the user ID of the affected user (not the actioning user)
> > currently, this is not returned in the view
> 

Not as clear re-reading (was late when I got done).  To be more accurate...

In the log entries of scope "user", the `entityId` field is not used anywhere in the vue views or code. It simply exists in the database.

> Can we be more specific about what 'typically' means here? Where are there exceptions to this?

In events like `users.create-user`, users.delete` where an admin is creating or deleting a user, the `UserID` field would be the logged in user (the one performing the action).  The `entityId` is the user being created/deleted/affected

In events like `account.reset-password`, user.update-user` where a user is is "affecting" themselves, the `UserID` field and `entityId` field would be the logged in user (self)

In events like `account.login` where the user ID is unknown due to failed login attempt, neither the `UserID` field or the `entityId` field would be set.

I am in the process of listing these out in the spreadsheet.

**Comment by Steve-Mcl** [comment]
> As I said in our conversation on Monday, I was not in favour of spending time now reworking the format of all of the audit log entries

At the time of discussion, only the `platform.x.y` entries were done - I did not intentionally rework these other that to move their scope to 'user' as discussed. However when I was reviewing code & preparing to commit last night, I could not bring myself to push the disjointed/irregular format. It certainly didnt add "days of work" 🙂  I assure you  👼 that was in adding the others & ironing out bugs, timing issues, figuring out how to (manually) test the fail cases etc etc

> and I'm not clear how much more work still needs to be done.

Me neither because as I stated, There are likely more but as there is no acceptance criteria I am having to hunt down the gaps.  

I would therefore propose we consider any missing log entries an iteration once this is merged?

**Comment by knolleary** [comment]
> Are you sure you want me to remove this? I aligned with the API reply and many (most) of them reply {status: 'okay'}
I simply included it as there can be value (i.e. something can complete but not be an error and not be ok).

To be frank, I want to review where our api returns that type of status response and consider if its really necessary. Most of our API returns data - and the presence of the data is all you need to know that the call worked.

But here we are not talking about API responses. They are logs of things happening. We do not need to fill the database with `{status:'okay'}`. For example, if it is an event that is only ever logged because it has happened, then including a status property is entirely redundant.

**Comment by knolleary** [comment]
> I am in the process of listing these out in the spreadsheet.

I think the main requirement I have is to be able to answer the question: 'show me all activity involving User X' - and for us to have a clear way to construct that query.

> I would therefore propose we consider any missing log entries an iteration once this is merged?

I think you have hit the main gaps for sure. The original issue wasn't exhaustive in listing the areas that needed covering, and part of this work was to identify the main gaps.

**Comment by Steve-Mcl** [comment]
@knolleary 

Nick, I've finished the spreadsheet (added some additional info I feel will be necessary for anyone consuming the events/log data)

Please have a quick peak.

<br>

I'll quickly strip out the `{ status: "okay" }` from the `body` and where nothing is left in the `body` data, I will send `null` instead of an empty object (as other log entries do)

**Comment by Steve-Mcl** [comment]
@knolleary this is now ready for review

**Review by knolleary — COMMENTED** [review]
A few comments based on a code review - I'm very conscious of minimising the data we are storing, especially around user email addresses as they are PII and need to be managed carefully. I can see some places where the logs appear to capture username, name and email where just username would do.

We should only be storing the email/name if it is absolutely necessary.

I now need to spend some time running it to see what the logs look like in practice and get a feel for it in terms of the volume of information being captured.

File: frontend/src/pages/admin/AuditLog.vue [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+import { mapState } from 'vuex'
+import SectionTopMenu from '@/components/SectionTopMenu'
+import AuditLog from '@/components/AuditLog'
+import permissionsMixin from '@/mixins/Permissions'
+import adminApi from '@/api/admin'
+
+export default {
+    name: 'PlatformAuditLog',
+    mixins: [permissionsMixin],
+    data () {
```
[the remark this answers is not in the document]
Comment (Steve-Mcl) [reply frontend/src/pages/admin/AuditLog.vue:16]: ah well spotted

File: frontend/src/mixins/Permissions.js [hunk 2]
```diff
@@ -7,6 +7,12 @@ export default {
                 throw new Error(`Unrecognised scope requested: '${scope}'`)
             }
             const permission = Permissions[scope]
+            if (scope.startsWith('platform:')) {
+                if (this.user?.admin) {
+                    return true
+                }
+                return false
+            }
```
[the remark this answers is not in the document]
Comment (Steve-Mcl) [reply frontend/src/mixins/Permissions.js:15]: yes, looks like I didnt fully remove the `platform:audit-log` permission I first added.

Comment (Steve-Mcl) [inline frontend/src/mixins/Permissions.js:15]:
```suggestion
```

remove unused code

[the remark this answers is not in the document]
Comment (Steve-Mcl) [reply frontend/src/mixins/Permissions.js]: resolved in [f82ad3d](https://github.com/flowforge/flowforge/pull/1084/commits/f82ad3d305a77c10ef2ba13ee75ac6ed402278db)

File: forge/routes/api/shared/users.js [hunk 3]
```diff
@@ -1,11 +1,39 @@
...
+     * @param {import("node_modules/fastify/fastify").FastifyReply} reply The HTTP reply object
+     * @param {userLogCallback} [userLog] Audit Log function
+     */
+    updateUser: async (app, user, request, reply, userLog) => {
+        const noop = async () => {}
+        userLog = userLog || noop
+        const logUserInfo = {
+            username: user.username,
+            name: user.name,
+            email: user.email
```
Comment (Steve-Mcl) [inline forge/routes/api/shared/users.js:33]:
```suggestion
            username: user.username,
```

File: forge/routes/api/admin.js [hunk 4]
```diff
@@ -52,9 +52,21 @@ module.exports = async function (app) {
...
+                    'platform.licence.apply',
+                    { license }
+                )
+                reply.send(license)
             } else if (request.body.action === 'inspect') {
-                reply.send(await app.license.inspect(request.body.license))
+                const license = await app.license.inspect(request.body.license)
+                await app.db.controllers.AuditLog.platformLog(
+                    request.session.User.id,
+                    'platform.licence.inspect',
```
[the remark this answers is not in the document]
Comment (Steve-Mcl) [reply forge/routes/api/admin.js:66]: Undecided. If someone was "testing out" a handfull of licences - would admins not want an audit trail?

File: forge/routes/api/shared/users.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (Steve-Mcl) [reply forge/routes/api/shared/users.js]: Resolved in [13b411d](https://github.com/flowforge/flowforge/pull/1084/commits/13b411df81a4566b569de6bdb9c572a38b9b043e)

File: forge/routes/api/admin.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (Steve-Mcl) [reply forge/routes/api/admin.js:66]: @knolleary I am pretty sure this the last point to address.  I am fine with keeping or dropping this particular entry. Either way, let me know.  Ta.

[the remark this answers is not in the document]
Comment (knolleary) [reply forge/routes/api/admin.js:66]: I wouldn't have added it. Not worth the effort to remove at this point.

File: forge/routes/api/shared/users.js [hunk 5]
```diff
@@ -1,11 +1,37 @@
...
+     * @param {object} user The user to update
+     * @param {import("node_modules/fastify/fastify").FastifyRequest} request The incoming request
+     * @param {import("node_modules/fastify/fastify").FastifyReply} reply The HTTP reply object
+     * @param {userLogCallback} [userLog] Audit Log function
+     */
+    updateUser: async (app, user, request, reply, userLog) => {
+        const noop = async () => {}
+        userLog = userLog || noop
+        const logUserInfo = {
+            username: user.username,
```
Comment (knolleary) [inline forge/routes/api/shared/users.js:31]:
```suggestion
            username: user.username
```

File: forge/routes/api/users.js [hunk 6]
```diff
@@ -79,15 +79,23 @@ module.exports = async function (app) {
             }
         }
     }, async (request, reply) => {
+        const logUserInfo = {
+            username: request.body.username,
+            admin: !!request.body.isAdmin,
```
Comment (knolleary) [inline forge/routes/api/users.js:84]:
```suggestion
            admin: !!request.body.isAdmin
```

**Review by knolleary — COMMENTED** [review]
A couple linting issues following the last set of changes - fixes proposed and I'll merge them.

Can you resolve the merge conflict?

**Comment by Steve-Mcl** [comment]
> Can you resolve the merge conflict?

resolved in 6ef418e (Merged main).

**Review by knolleary — APPROVED** [review]

**Pull request merged**